src/hotspot/share/opto/loopopts.cpp

```diff
@@ -4120,3 +4121,204 @@ bool PhaseIdealLoop::duplicate_loop_backedge(IdealLoopTree *loop, Node_List &old
+    uint vector_length    = vec_t->length();
+    BasicType bt          = vec_t->element_basic_type();
+    const Type* bt_t      = Type::get_const_basic_type(bt);
+
+    // Create vector of identity elements (zero for add, one for mul, etc)
+    Node* identity_scalar = ReductionNode::make_identity_input_for_reduction_from_vector_opc(_igvn, last_ur->Opcode(), bt);
+    set_ctrl(identity_scalar, C->root());
+    VectorNode* identity_vector = VectorNode::scalar2vector(identity_scalar, vector_length, bt_t);
+    _igvn.register_new_node_with_optimizer(identity_vector);
+    set_ctrl(identity_vector, C->root());
+    assert(vec_t == identity_vector->vect_type(), "matching vector type");
+#ifdef ASSERT
+    if (TraceNewVectors) {
+      tty->print("new Vector node: ");
+      identity_vector->dump();
+    }
+#endif
+
+    // Traverse down the chain of UnorderedReductions, and create the vector_accumulators.
+    current = first_ur;
+    Node* last_vector_accumulator = phi;
+    while (true) {
+      // Create vector_accumulator to replace current.
+      Node* vector_input = current->in(2);
+      VectorNode* vector_accumulator = current->make_normal_vector_op(last_vector_accumulator, vector_input, vec_t);
+      _igvn.register_new_node_with_optimizer(vector_accumulator);
+      C->copy_node_notes_to(vector_accumulator, current);
+      set_ctrl(vector_accumulator, cl);
+      last_vector_accumulator = vector_accumulator;
+      if (current != last_ur) {
+        // All UnorderedReductions except the last are now useless.
+        _igvn.rehash_node_delayed(current);
+        current->set_req_X(1, C->top(), &_igvn);
+        current->set_req_X(2, C->top(), &_igvn);
+      }
+#ifdef ASSERT
+      if (TraceNewVectors) {
+        tty->print("new Vector node: ");
+        vector_accumulator->dump();
+      }
+#endif
+      // Iterate down, until we hit last_ur.
+      if (current != last_ur) {
+        current = current->unique_out()->as_UnorderedReduction();
+      } else {
+        current = nullptr;
+        break;
+      }
+    }
+    assert(current == nullptr, "terminated correctly");
+
+    // After the loop, we can reduce the init and last_vector_accumulator.
+    Node* init = phi->in(1);
+    _igvn.rehash_node_delayed(last_ur);
+    last_ur->set_req_X(1, init, &_igvn);
+    last_ur->set_req_X(2, last_vector_accumulator, &_igvn);
+
+    // Turn the scalar phi into a vector phi.
+    _igvn.rehash_node_delayed(phi);
+    phi->set_req_X(1, identity_vector, &_igvn);
+    phi->set_req_X(2, last_vector_accumulator, &_igvn);
+    phi->as_Type()->set_type(vec_t);
+    _igvn.set_type(phi, vec_t);
+    assert(phi->outcnt() == 1, "accumulator is only use of phi");
+
+    // Update control to outside the loop.
+    Node* new_ctrl = get_late_ctrl(last_ur, cl);
+    set_ctrl(last_ur, new_ctrl);
+    assert(new_ctrl != nullptr && new_ctrl != cl, "new control of ur must be outside loop");
+
+#ifdef ASSERT
+    if (TraceNewVectors) {
```

Comment:
Not sure if this is required. But I haven't seen the creation of vector phi is traced in other places.

Reply:
Do we even have any other Vector Phi's? SuperWord never does it. Maybe in the Vector API?

Reply:
My colleague working on VectorAPI just told me yes. A scalar phi may eventually become a vector phi after several transformations in vector intrinsics and vector box/unbox optimizations.

Reply:
Ok, I can remove it. The user probably does not care about that so much, more about the vector instructions that actually end up in the code-gen.